tests/EdgeToEdge/APIRoutes/UpgradeMembershipFeeRouteTest.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace EdgeToEdge\APIRoutes;
+
+use PHPUnit\Framework\Attributes\CoversClass;
+use PHPUnit\Framework\Attributes\DataProvider;
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use WMDE\Fundraising\Frontend\App\Controllers\API\Membership\UpgradeMembershipFeeController;
+use WMDE\Fundraising\Frontend\Tests\EdgeToEdge\WebRouteTestCase;
+use WMDE\Fundraising\Frontend\Tests\Fixtures\LoggerSpy;
+use WMDE\Fundraising\MembershipContext\Tests\Fixtures\FeeChanges;
+use WMDE\Fundraising\PaymentContext\Domain\PaymentType;
+
+#[CoversClass( UpgradeMembershipFeeController::class )]
+class UpgradeMembershipFeeRouteTest extends WebRouteTestCase {
+
+	private LoggerSpy $loggerSpy;
+	private KernelBrowser $client;
+
+	public function setUp(): void {
+		$this->client = $this->createClient();
+
+		$this->loggerSpy = new LoggerSpy();
+		$this->getFactory()->setLogger( $this->loggerSpy );
+	}
+
+	public function testControllerReceivesIncompleteRequestData_ThrowsInternalExceptionDueToParsingErrors(): void {
+		$this->client->jsonRequest(
+			method:'POST',
+			uri:'/api/v1/membership/change-fee',
+			parameters: []
+		);
+
+		$this->loggerSpy->assertNoLoggingCallsWhereMade();
+
+		$response = $this->client->getResponse();
+		$this->assertEquals(
+			 [ "ERR" => "This value should be of type string.\nThis value should be of type int.\nThis value should be of type string." ],
+			json_decode( $response->getContent() ?: '', true )
+		);
```

Comment:
This message will change whenever `FeeChangeRequest` changes, which makes the test brittle. I'd say you can check for an `ERR` key in the response, but not check the exact contents.

I'll create a pull request that'll make the handled errors for the content validation a bit more informative for the developers (adding the field name to the type violation), but that behavior does not need to be tested in this test.